### PR TITLE
SRE-1055 rvm: command not found

### PR DESCRIPTION
**Pull request by whatisdot**
Closes https://snapsheettech.atlassian.net/browse/SRE-1055

Builds were passing when we tested them via docker compose using Linux, but the process that was used for installation on Linux did not align properly with MacOS. This fixes installations for both operating systems by creating an isolated version of Ruby that won't overlap with any existing installation of Ruby. It uses RVM to achieve this (see code documentation comments).

### `DebugTools`
A new module that can be shared between formulas and download strategies for enabling `binding.pry` breakpoints when attempting to install a formula with `--debug`.

### `RubyManager`
This manages the building and installation of a dedicated version of Ruby using RVM. This requires that the user has RVM installed, and that RVM is installed using the single-user method, which is recommended in their documentation. Because of this, I have modified the Dockerfile so the Linux container will also install RVM in user space.

### `RubyGemsDownloadStrategy`
This module was using the Homebrew-native version of Ruby to download the gem, which could cause issues if/when Homebrew makes changes to this the next time a user does a `brew update && brew upgrade`. By using the $SHELL of the user's environment, the version of `gem` used will be from the user's environment.

### `Tinker`
This as been refactored to reflect the changes above.

### Testing Linux

Checkout this branch and build the docker image for development.
```
docker compose build
```

**Known Issue**: Attempting to run the installer in Linux more than once within the same container will show an error. This can be resolved terminating the container and starting a new one. This does not occur on MacOS.

Attempt to install Tinker on the docker image by starting a container:
```
docker compose run --rm cli
```

Within the container, try the installer with the `--debug` option, and confirm Tinker was installed:
```
brew install --debug --build-from-source Formula/tinker.rb
tinker --version
```

Stop the container and start a new one. Then confirm you can install without the `--debug` option:
```
brew install --build-from-source Formula/tinker.rb
tinker --version
```

### Testing MacOS

Checkout this branch and make sure you don't have tinker already installed:
```
brew remove tinker
```

Install Tinker with the `--debug` flag. This assumes that you have `GITHUB_TOKEN` set on your environment. This will take about 9min...
```
HOMEBREW_GITHUB_API_TOKEN=$GITHUB_TOKEN brew install --debug --build-from-source Formula/tinker.rb
tinker --version
```

Remove Tinker and confirm that you can install without the `--debug` flag, as most users will. This will also take about 9min...
```
HOMEBREW_GITHUB_API_TOKEN=$GITHUB_TOKEN brew install --build-from-source Formula/tinker.rb
tinker --version
```

### CODE REVIEW

**Review by d10zero — CHANGES_REQUESTED** [review]
Downloading in the docker container worked as expected. but when testing MacOS, I see the error below. I also get an error when I initially tried running `brew remove tinker`

```
 ✔  brew remove tinker
Error: No such keg: /usr/local/Cellar/tinker
```

```
 ↵ 127 
HOMEBREW_GITHUB_API_TOKEN=<token> brew install --debug --build-from-source Formula/tinker.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FromPathLoader): loading /Users/deena.tenzer/Documents/homebrew-core/Formula/tinker.rb
Error: Failed to load cask: Formula/tinker.rb
Cask 'tinker' is unreadable: wrong constant name #<Class:0x00007f8b22d51510>
Warning: Treating Formula/tinker.rb as a formula.
==> Fetching tinker gem from GitHub packages.
Downloaded tinker-0.2.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/snapsheet/homebrew-core/Formula/tinker.rb
==> Installing dedicated version of Ruby 2.7.1. This will take several minutes.
==> For detailed progress information, run the following in a new CLI:
==>   tail -f /Users/deena.tenzer/Library/Logs/Homebrew/tinker/01.rvm
==> /usr/local/Cellar/tinker/0.2.1/.rvm/bin/rvm --autolibs=read-fail --with-gcc=gcc --verbose --debug install 2.7.1
Last 15 lines from /Users/deena.tenzer/Library/Logs/Homebrew/tinker/01.rvm:
__rvm_setup_compile_environment_movable_early ruby-2.7.1
__rvm_setup_compile_environment_system_early ruby-2.7.1
__rvm_setup_compile_environment_requirements ruby-2.7.1
Checking requirements for osx.
requirements code for osx loaded
requirements lib type set to osx_brew
Do not know how to check/install gcc 'gcc'.
Found required packages: autoconf, automake, libtool, pkg-config, coreutils, libyaml, libksba, readline, zlib, openssl@1.1.
Homebrew libs 'automake coreutils libtool openssl@1.1 readline' require update - skipping.
RVM autolibs is now configured with mode '2' =>
  'Allow RVM to use package manager if found, fail if dependencies are missing. This is default.',
please run `rvm autolibs enable` to let RVM do its job or run and read `rvm autolibs [help]`
or visit https://rvm.io/rvm/autolibs for more information.
Requirements installation failed with status: 1.
__rvm_rm_rf already gone: /usr/local/Cellar/tinker/0.2.1/.rvm/tmp/84209*
/usr/local/Homebrew/Library/Homebrew/shims/shared/git --version
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --version
/usr/local/Homebrew/Library/Homebrew/ignorable.rb:29:in `block in raise'
BuildError: Failed executing: /usr/local/Cellar/tinker/0.2.1/.rvm/bin/rvm --autolibs=read-fail --with-gcc=gcc --verbose --debug install 2.7.1
1. raise
2. ignore
3. backtrace
4. irb
5. shell
Choose an action: irb
When you exit this IRB session, execution will continue.
irb(tinker):001:0> exit
==> Installing ruby gem for tinker 0.2.1.
==> For detailed progress information, run the following in a new CLI:
==>   tail -f /Users/deena.tenzer/Library/Logs/Homebrew/tinker/02.gem
==> gem install /Users/deena.tenzer/Library/Caches/Homebrew/tinker-0.2.1.gem
Last 15 lines from /Users/deena.tenzer/Library/Logs/Homebrew/tinker/02.gem:
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:361:in `block in postpone'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:331:in `open'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:357:in `postpone'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:958:in `checking_for'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:1011:in `have_library'
	from extconf.rb:137:in `<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /usr/local/Cellar/tinker/0.2.1/.rvm/gems/ruby-2.7.1/extensions/universal-darwin-20/2.6.0/rugged-1.1.1/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /usr/local/Cellar/tinker/0.2.1/.rvm/gems/ruby-2.7.1/gems/rugged-1.1.1 for inspection.
Results logged to /usr/local/Cellar/tinker/0.2.1/.rvm/gems/ruby-2.7.1/extensions/universal-darwin-20/2.6.0/rugged-1.1.1/gem_make.out
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --version
/usr/local/Homebrew/Library/Homebrew/ignorable.rb:29:in `block in raise'
BuildError: Failed executing: gem install /Users/deena.tenzer/Library/Caches/Homebrew/tinker-0.2.1.gem
1. raise
2. ignore
3. backtrace
4. irb
5. shell
Choose an action: ignore
/Users/deena.tenzer/Documents/homebrew-core/lib/ruby_gems_download_strategy.rb:31:in `unlink'
Errno::ENOENT: No such file or directory @ apply2files - /private/tmp/.gemrc
1. raise
2. backtrace
3. shell
Choose an action: shell
When you exit this shell, you will return to the menu.
deena.tenzer@deena tinker-20220125-84185-a3b91b % exit
1. raise
2. backtrace
3. shell
Choose an action: raise
==> Temporary files retained at:
/private/tmp/tinker-20220125-84185-a3b91b
Error: An exception occurred within a child process:
  Errno::ENOENT: No such file or directory @ apply2files - /private/tmp/.gemrc
/Users/deena.tenzer/Documents/homebrew-core/lib/ruby_gems_download_strategy.rb:31:in `unlink'
/Users/deena.tenzer/Documents/homebrew-core/lib/ruby_gems_download_strategy.rb:31:in `gem_config_file'
/Users/deena.tenzer/Documents/homebrew-core/Formula/tinker.rb:74:in `install_tinker_gem'
/Users/deena.tenzer/Documents/homebrew-core/Formula/tinker.rb:121:in `install'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:17:in `block in install'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:90:in `debrew'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:17:in `install'
/usr/local/Homebrew/Library/Homebrew/build.rb:172:in `block (3 levels) in install'
/usr/local/Homebrew/Library/Homebrew/utils.rb:588:in `with_env'
/usr/local/Homebrew/Library/Homebrew/build.rb:134:in `block (2 levels) in install'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1296:in `block in brew'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2475:in `block (2 levels) in stage'
/usr/local/Homebrew/Library/Homebrew/utils.rb:588:in `with_env'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2474:in `block in stage'
/usr/local/Homebrew/Library/Homebrew/resource.rb:126:in `block (2 levels) in unpack'
/usr/local/Homebrew/Library/Homebrew/download_strategy.rb:117:in `chdir'
/usr/local/Homebrew/Library/Homebrew/download_strategy.rb:102:in `stage'
/usr/local/Homebrew/Library/Homebrew/resource.rb:122:in `block in unpack'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `block in run'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `chdir'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `run'
/usr/local/Homebrew/Library/Homebrew/resource.rb:208:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/resource.rb:121:in `unpack'
/usr/local/Homebrew/Library/Homebrew/resource.rb:96:in `stage'
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/lib/ruby/2.6.0/forwardable.rb:230:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2454:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1289:in `brew'
/usr/local/Homebrew/Library/Homebrew/build.rb:129:in `block in install'
/usr/local/Homebrew/Library/Homebrew/utils.rb:588:in `with_env'
/usr/local/Homebrew/Library/Homebrew/build.rb:124:in `install'
/usr/local/Homebrew/Library/Homebrew/build.rb:224:in `<main>'
```

File: Dockerfile [hunk 1]
```diff
@@ -25,17 +25,10 @@ RUN sudo apt-get update \
     cmake \
   && curl -sSL https://rvm.io/mpapis.asc | gpg --import -
 
-# install RVM's ordered dependencies and repos
-RUN sudo apt-get install software-properties-common \
-  && sudo apt-add-repository -y ppa:rael-gc/rvm \
-  && sudo apt-get update \
-  && sudo apt-get install -y libfile-fcntllock-perl
+USER dev.user
```
Comment (deekom) [inline Dockerfile:28]: Does the user context change between line 14 and line 28, that necessitates resetting `USER dev.user`

**Review by deekom — APPROVED** [review]
Was able to install tinker on MacOS as well as inside linux container(both with and without debug) with no errors.
Screen Shot 2022-01-26 at 12 08 59 PM

**Review by d10zero — CHANGES_REQUESTED** [review]
I am not getting the following error:

```
 HOMEBREW_GITHUB_API_TOKEN=<token> brew install --debug --build-from-source Formula/tinker.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FromPathLoader): loading /Users/deena.tenzer/Documents/homebrew-core/Formula/tinker.rb
Error: Failed to load cask: Formula/tinker.rb
Cask 'tinker' is unreadable: wrong constant name #<Class:0x00007f865b0dc468>
Warning: Treating Formula/tinker.rb as a formula.
==> Fetching tinker gem from GitHub packages.
Error: cannot load such file -- interception
Please report this issue:
  https://docs.brew.sh/Troubleshooting
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-rescue-1.5.2/lib/pry-rescue.rb:2:in `require'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-rescue-1.5.2/lib/pry-rescue.rb:2:in `<top (required)>'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-rescue-1.5.2/lib/pry-rescue/cli.rb:1:in `require'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-rescue-1.5.2/lib/pry-rescue/cli.rb:1:in `<top (required)>'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-0.13.1/lib/pry/plugins.rb:55:in `require'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-0.13.1/lib/pry/plugins.rb:55:in `load_cli_options'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-0.13.1/lib/pry/cli.rb:40:in `each'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-0.13.1/lib/pry/cli.rb:40:in `add_plugin_options'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-0.13.1/lib/pry/cli.rb:134:in `<top (required)>'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-0.13.1/lib/pry.rb:78:in `require'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-0.13.1/lib/pry.rb:78:in `<top (required)>'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-byebug-3.9.0/lib/pry-byebug.rb:3:in `require'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/pry-byebug-3.9.0/lib/pry-byebug.rb:3:in `<top (required)>'
/Users/deena.tenzer/Documents/homebrew-core/lib/debug_tools.rb:5:in `require'
/Users/deena.tenzer/Documents/homebrew-core/lib/debug_tools.rb:5:in `setup_debug_tools'
/Users/deena.tenzer/Documents/homebrew-core/lib/ruby_gems_download_strategy.rb:53:in `fetch'
/usr/local/Homebrew/Library/Homebrew/resource.rb:147:in `fetch'
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/lib/ruby/2.6.0/forwardable.rb:230:in `fetch'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2076:in `fetch'
/usr/local/Homebrew/Library/Homebrew/formula_installer.rb:1187:in `fetch'
/usr/local/Homebrew/Library/Homebrew/install.rb:305:in `block in install_formulae'
/usr/local/Homebrew/Library/Homebrew/install.rb:278:in `map'
/usr/local/Homebrew/Library/Homebrew/install.rb:278:in `install_formulae'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:218:in `install'
/usr/local/Homebrew/Library/Homebrew/brew.rb:110:in `<main>'
```

**Review by d10zero — APPROVED** [review]
Was able to install Tinker in linux as well as mac successfully

**Pull request merged**